### DISCUSSION TITLE
Rate-limiting for server API

### DISCUSSION THREAD

**Discussion by lukehutch** [serverpod/serverpod#1860] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**

The Serverpod server needs API rate-limiting, in particular to prevent databases from being mined by determined hackers.

**Describe the solution you'd like**

The server needs to honor a developer-provided rate-limit for incoming API calls (this could also be applied to the built-in web server, but I'm more interested in rate-limiting the API endpoints).

I think Redis allows a developer to set a time to live for a record, before it is expunged, which would provide a nice way for rate limit data to be tracked without the data growing without bound.

Ideally I would like to be able to rate-limit based on both IP address and userId, but to rate-limit based on userId, first support for JWTs would need to be added (#1720).

**Describe alternatives you've considered**

Writing a rate limiter that I can manually call from each endpoint (not an ideal situation -- this should be built into Serverpod).

**Additional context**

I know this is a problem because I once mined the entire database of a major dating app, by installing a MitM proxy to steal my own login cookie, and then simply downloading the profile data and photos with curl requests. This was made a lot easier because the server assigned sequential userIds (#554) and did not implement certificate pinning (#702), but also because the developers did not implement any sort of rate-limiting (hence this issue). In fact, I opened something like 16 threads, and fetched many profiles at a time in parallel, and there was nothing to stop me! The only indication the admins got that I had mined their entire database was the spike in traffic.

There are many articles online about how people have done exactly this for different dating app backends. This is inevitably going to happen to any app that gets popular enough, especially for dating apps, like the one I am developing!

**Comment by Isakdl** [comment, marked as answer]
You can achieve this today on the load balancer level, for example Cloud Armor on GCP https://cloud.google.com/armor/docs/rate-limiting-overview.

Here is an example of how to set it up with HTTP(s) Load balancer https://medium.com/google-cloud/how-to-deploy-google-cloud-armor-with-http-s-load-balancer-using-terraform-245805813d3c.

There are of course similar solutions on other platforms. I tend to lean towards having this stuff at the infrastructure level rather than the application layer, as this leaves out a lot of complexities from the server. You also get the benefit that if you intercept it earlier you can easily deal with traffic hitting multiple different instances without having to think about syncing this between them. On top of that the servers will never have to expend CPU time or do any work as the requests will not even hit them.

While dealing with infra is not trivial I think it provides the best solution.

**Comment by lukehutch** [comment]
@Isakdl Got it. Thanks for the links. Should I close this issue?

**Comment by vanlooverenkoen** [comment]
@Isakdl Something I am still missing, I am hosting everything on a linux server provided by the client. So implementing rate limiting is quite hard in this case I think.

It would be nice to have middleware that we can add to every request to do some kind of security checks. 
`onRequest`
`afterResponse`

Just to have some triggers where we can do some basic checks, it would also simplify a custom implementation of `requireLogin` to have our own validation checks here.

**Comment by macMikey** [comment]
[Rodrigo Mansueli](https://hashnode.com/@mansueli), who is a supabase support engineer, [wrote a blog post about doing this just within postgres](https://blog.mansueli.com/rate-limiting-supabase-requests-with-postgresql-and-pgheaderkit)
